[JSC] ASSERTION FAILED: child(0)->type() == phi->type()

hyjorc1 opened the pull request:
#### 7d6c85ebcb93c10e4c31cbdae6bcadbdd11e2158
<pre>
[JSC] ASSERTION FAILED: child(0)-&gt;type() == phi-&gt;type()
<a href="https://rdar.apple.com/147364146">rdar://147364146</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290009">https://bugs.webkit.org/show_bug.cgi?id=290009</a>

Reviewed by Yusuke Suzuki.

Array allocation sinking introduced cases where Upsilon edges may come from nodes
with result types like double or int52, instead of the previously assumed JS result
type. This breaks assumptions made during Phi construction, which always expected
NodeResultJS.

This patch adds `fixUpsilonEdge()`, a pass that ensures Upsilon inputs match the
expected Phi result type by inserting appropriate ValueRep nodes when needed.

* JSTests/stress/array-allocation-sink-upsilon-with-double-value.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGFlushFormat.h:

Canonical link: <a href="https://commits.webkit.org/292572@main">https://commits.webkit.org/292572@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2abfcb8c60f6bf91987e66c2e21f37df0e0a835c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101362 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46814 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24342 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73394 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30622 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99298 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86981 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53731 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/95782 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4748 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46141 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88968 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4846 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103390 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94916 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17005 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82433 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23613 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83006 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81811 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/20570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26447 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3881 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16736 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15528 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23325 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28480 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118393 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22984 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33418 "Found 5 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.mini-mode, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/try-catch-loop-osr.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26464 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGFlushFormat.h:
```diff
@@ -67,6 +67,29 @@ inline NodeFlags resultFor(FlushFormat format)
+    }
+    return 0;
+}
+
```
I think this looks a bit too fragile since we have no guarantee that any other UseKind will come here.

hyjorc1 replied:
If you check the other places where we insert `Upsilon` node. See `DFGSSAConversionPhase.cpp` and `DFGPutStackSinkingPhase.cpp` the use kind is from `uncheckedUseKindFor`
```
UseKind useKind = uncheckedUseKindFor(format);

dataLogLnIf(verbose, "Inserting Upsilon for ", variable->operand(), " propagating ", valueForOperand.operand(variable->operand()), " to ", phiNode);

m_insertionSet.insertNode(
    upsilonInsertionPoint, SpecNone, Upsilon, upsilonOrigin,
    OpInfo(phiNode), Edge(
        valueForOperand.operand(variable->operand()),
        useKind));
```

and for the object allocation sinking phase, the Upsilon edge use kind is from node's `defaultEdge` which is based on `useKindForResult`.

```
UseKind defaultUseKind()
    {
        return useKindForResult(result());
    }
    
    Edge defaultEdge()
    {
        return Edge(this, defaultUseKind());
    }
```

After we see the implementation of `useKindForResult`, `resultFor(FlushFormat format)`, `useKindFor(FlushFormat format)`, and `uncheckedUseKindFor(FlushFormat format)`. We can figure out that the they (`NodeFlags` and `UseKind` for `Upsilon` and `Phi`) are strictly related to the `FlushFormat`. So, the added `resultFor(UseKind usekind)` should make sense since `resultFor(UseKind usekind)` is derived from `resultFor(FlushFormat format)` and `useKindFor(FlushFormat format)`.

hyjorc1 replied:
removed since it's not needed in this fix.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2819,6 +2820,47 @@ class ObjectAllocationSinkingPhase : public Phase {
+
+                        edge.setNode(result);
+                        edge.setUseKind(useKindForResult(NodeResultJS));
+                    }
```
To make this more robust, I wonder if we should avoid using `resultFor`.
Like,
```
if (node->phi()->hasJSResult()) {
    Node* result = nullptr;
    if (edge->hasDoubleResult())
        result = m_insertionSet.insertNode(indexInBlock, SpecBytecodeDouble, ValueRep, node->origin, Edge(edge.node(), DoubleRepUse));
    else if (edge->hasInt52Result())
        result = ...

    if (result) {
        edge.setNode(result);
        edge.setUseKind(UntypedUse);
    }
}
```

hyjorc1 replied:
Will change. For these use kinds they must be 
```
inline UseKind useKindForResult(NodeFlags result)
{
    ASSERT(!(result & ~NodeResultMask));
    switch (result) {
    case NodeResultInt52:
        return Int52RepUse;
    case NodeResultDouble:
        return DoubleRepUse;
    default:
        return UntypedUse;
    }
}
```
since we use `defaultEdge()` for inserting Upsilon nodes in `DFGObjectAllocationSinkingPhase.cpp`.
```
m_insertionSet.insertNode(
                        upsilonInsertionPoint, SpecNone, Upsilon, upsilonOrigin,
                        OpInfo(phiNode), incoming->defaultEdge());
```

hyjorc1 replied:
Updated.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] ASSERTION FAILED: child(0)->type() == phi->type()
rdar://147364146
https://bugs.webkit.org/show_bug.cgi?id=290009

Reviewed by Yusuke Suzuki.

Array allocation sinking introduced cases where Upsilon edges may come from nodes
with result types like double or int52, instead of the previously assumed JS result
type. This breaks assumptions made during Phi construction, which always expected
NodeResultJS.

This patch adds `fixUpsilonEdge()`, a pass that ensures Upsilon inputs match the
expected Phi result type by inserting appropriate ValueRep nodes when needed.

* JSTests/stress/array-allocation-sink-upsilon-with-double-value.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGFlushFormat.h:

Canonical link: https://commits.webkit.org/292572@main
